components/layout/query.rs

```diff
@@ -527,21 +531,69 @@ impl FragmentBorderBoxIterator for UnioningFragmentScrollAreaIterator {
+        }
+
         if fragment.node == self.node_address {
             // Found the fragment in the flow tree that matches the
             // DOM node being looked for.
-            self.has_found_node = true;
-            self.node_border_box = *border_box;
+
+            // If self.node_offset_box is Some, we were treating this as an inline node!
+            assert!(self.node_offset_box.is_none());
+
+            self.has_processed_node = true;
+            self.node_offset_box = Some(NodeOffsetBoxInfo {
+                offset: border_box.origin,
+                rectangle: *border_box,
+            });
 
             // offsetParent returns null if the node is fixed.
             if fragment.style.get_box().position == computed_values::position::T::fixed {
                 self.parent_nodes.clear();
             }
-        } else if level > self.last_level {
+        } else if let Some(node) = fragment.inline_context.as_ref().and_then(|inline_context| {
+            inline_context.nodes.iter().find(|node| node.address == self.node_address)
+        }) {
+            // TODO: Handle cases where the `offsetParent` is an inline
+            // element. This will likely be impossible until
+            // https://github.com/servo/servo/issues/13982 is fixed.
+
+            // Found a fragment in the flow tree whose inline context contains
+            // the DOM node we're looking for, i.e. the node is inline and
+            // contains this fragment.
+            match self.node_offset_box {
+                Some(NodeOffsetBoxInfo { ref mut rectangle, .. }) => {
+                    *rectangle = rectangle.union(border_box);
+                },
+                None => {
+                    // https://github.com/servo/servo/issues/13982 will cause
+                    // this assertion to fail sometimes, so it's commented out
+                    // for now.
+                    //assert!(node.flags.contains(FIRST_FRAGMENT_OF_ELEMENT));
+
+                    self.node_offset_box = Some(NodeOffsetBoxInfo {
+                        offset: border_box.origin,
+                        rectangle: *border_box,
+                    });
+                },
+            }
+
+            if node.flags.contains(LAST_FRAGMENT_OF_ELEMENT) {
+                self.has_processed_node = true;
+            }
+
+            // TODO: `position: fixed` on inline elements doesn't seem to work
+            // as of Sep 25, 2016, so I don't know how one will check if an
+            // inline element has it when it does.
+        } else if self.node_offset_box.is_none() {
             // TODO(gw): Is there a less fragile way of checking whether this
             // fragment is the body element, rather than just checking that
-            // the parent nodes stack contains the root node only?
-            let is_body_element = self.parent_nodes.len() == 1;
+            // it's at level 1 (above the root node)?
```

Comment:
does this comment intend to say "just below the root node"? (I think the usual visualization is with the root at the top)